### ISSUE TITLE
Invalid URL error with some encoded and plain URLs

### ISSUE THREAD

**Issue by cesarhdz** [DarthSim/imgproxy#129]
I'm using docker in ec2 to serve img proxy and I'm not able to use in my project because some image URLs doesn't work. Sometimes fails the plain version and other times the encoded version:

## 1. Some images work in plain format URL but not encoded

Encoded images ending with `=` charachter doesn't seem to work, I get _Invalid url encoding_ error, example: 

```
2019/01/14 02:15:18 [Qy_valDilLA_2VRxmxQoRl] GET: /insecure/s:180:92/aHR0cHM6Ly9zMy11cy13ZXN0LTIuYW1hem9uYXdzLmNvbS9ib29rYW5kbGVhcm4yZGV2ZWxvcC9pbWcvZWR1cGFzcy1sb2dvLmpwZw==
2019/01/14 02:15:18 | 404 | [Qy_valDilLA_2VRxmxQoRl] Invalid url encoding
2019/01/14 02:16:04 [Au5dAzYtw7Mig_sCAzlhxY] GET: /insecure/s:180:92/aHR0cHM6Ly9zMy11cy13ZXN0LTIuYW1hem9uYXdzLmNvbS9ib29rYW5kbGVhcm4yZGV2ZWxvcC9pbWcvZWR1cGFzcy1sb2dvLmpwZw==.jpg
2019/01/14 02:16:04 | 404 | [Au5dAzYtw7Mig_sCAzlhxY] Invalid url encoding
```

Y tried with and without extension, and none worked. But if I use the plain URL it works: 

```
2019/01/14 01:35:08 [4u4H-d9riTahDgnVy6B4zZ] GET: /insecure/s:180:92/plain/https://s3-us-west-2.amazonaws.com/bookandlearn2develop/img/edupass-logo.jpg
2019/01/14 01:35:08 | 200 | [4u4H-d9riTahDgnVy6B4zZ] Processed in 43.595298ms: https://s3-us-west-2.amazonaws.com/bookandlearn2develop/img/edupass-logo.jpg; &{Resize:fit Width:180 Height:92 Dpr:1 Gravity:{Type:ce X:0 Y:0} Enlarge:false Format:jpeg Quality:80 Flatten:false Background:{R:255 G:255 B:255} Blur:0 Sharpen:0 CacheBuster: Watermark:{Enabled:false Opacity:1 Replicate:false Gravity:ce OffsetX:0 OffsetY:0 Scale:0} UsedPresets:[]}
2019/01/14 01:17:34 [f-_RqDPstzWZEADAd3NezX] GET: /insecure/s:180:92/plain/https://s3-us-west-2.amazonaws.com/bookandlearn2develop/img/edupass-logo.jpg@jpg                                                                              
2019/01/14 01:17:34 | 200 | [f-_RqDPstzWZEADAd3NezX] Processed in 54.890664ms: https://s3-us-west-2.amazonaws.com/bookandlearn2develop/img/edupass-logo.jpg; &{Resize:fit Width:180 Height:92 Dpr:1 Gravity:{Type:ce X:0 Y:0} Enlarge:false Format:jpeg Quality:80 Flatten:false Background:{R:255 G:255 B:255} Blur:0 Sharpen:0 CacheBuster: Watermark:{Enabled:false Opacity:1 Replicate:false Gravity:ce OffsetX:0 OffsetY:0 Scale:0} UsedPresets:[]}
```

## 2. Some images work only with encoded URLs

Image containing `@` symbol returns _Invalid Image format_

```
2019/01/14 02:03:25 [rmoPV0BqJYPjQo_6mBHrMs] GET: /insecure/s:180:92/plain/https://bookandlearn.s3.amazonaws.com/distributor/3-icon@55bc37172eae4ccfb3569877dccef58d.jpg
2019/01/14 02:03:25 | 404 | [rmoPV0BqJYPjQo_6mBHrMs] Invalid image format: 55bc37172eae4ccfb3569877dccef58d.jpg
```

Image replacing `@` with `%40` doesn't work either

```
2019/01/14 02:04:38 [wZ-8a2DLC9mju1ooNZ0bYB] GET: /insecure/s:180:92/plain/https://bookandlearn.s3.amazonaws.com/distributor/3-icon%4055bc37172eae4ccfb3569877dccef58d.jpg
2019/01/14 02:04:38 | 404 | [wZ-8a2DLC9mju1ooNZ0bYB] Invalid image format: 55bc37172eae4ccfb3569877dccef58d.jpg
``` 

But images with encoded URL work

```
2019/01/14 01:45:02 [Btk0Jkyz66fY8bm_7o2Npz] GET: /insecure/s:180:92/aHR0cHM6Ly9ib29rYW5kbGVhcm4uczMuYW1hem9uYXdzLmNvbS9kaXN0cmlidXRvci8zLWljb25ANTViYzM3MTcyZWFlNGNjZmIzNTY5ODc3ZGNjZWY1OGQuanBn
2019/01/14 01:45:03 | 200 | [Btk0Jkyz66fY8bm_7o2Npz] Processed in 167.910188ms: https://bookandlearn.s3.amazonaws.com/distributor/3-icon@55bc37172eae4ccfb3569877dccef58d.jpg; &{Resize:fit Width:180 Height:92 Dpr:1 Gravity:{Type:ce X:0 Y:0} Enlarge:false Format:jpeg Quality:80 Flatten:false Background:{R:255 G:255 B:255} Blur:0 Sharpen:0 CacheBuster: Watermark:{Enabled:false Opacity:1 Replicate:false Gravity:ce OffsetX:0 OffsetY:0 Scale:0} UsedPresets:[]}
2019/01/14 01:46:02 [e8V6iMNFENqBCCg5bXU9C2] GET: /insecure/s:180:92/aHR0cHM6Ly9ib29rYW5kbGVhcm4uczMuYW1hem9uYXdzLmNvbS9kaXN0cmlidXRvci8zLWljb25ANTViYzM3MTcyZWFlNGNjZmIzNTY5ODc3ZGNjZWY1OGQuanBn.jpg
2019/01/14 01:46:02 | 200 | [e8V6iMNFENqBCCg5bXU9C2] Processed in 80.508922ms: https://bookandlearn.s3.amazonaws.com/distributor/3-icon@55bc37172eae4ccfb3569877dccef58d.jpg; &{Resize:fit Width:180 Height:92 Dpr:1 Gravity:{Type:ce X:0 Y:0} Enlarge:false Format:jpeg Quality:80 Flatten:false Background:{R:255 G:255 B:255} Blur:0 Sharpen:0 CacheBuster: Watermark:{Enabled:false Opacity:1 Replicate:false Gravity:ce OffsetX:0 OffsetY:0 Scale:0} UsedPresets:[]}
```

**Comment by DarthSim** [comment]
Hi,

It's expected behavior:

1. All `=` at the end of encoded URL should be removed. I think I'll trim 'em inside of imgproxy, but currently, they should be trimmed manually.

2. If a source URL contains a query string or `@`, the whole source URL should be escaped. In your case, it should be `https%3A%2F%2Fbookandlearn.s3.amazonaws.com%2Fdistributor%2F3-icon%4055bc37172eae4ccfb3569877dccef58d.jpg`.

**Comment by cesarhdz** [comment]
Hi @DarthSim , thanks for your quick reply!

I trimmed the `=` and it worked, all images are procesed correctly. I tried escaping the URL but it didn't seem to work. Anyway, I think the issue is solved since I can use the encoded URL.

**Comment by DarthSim** [comment]
I released v2.1.5 that removes `=` internally.

**Comment by Grawl** [comment]
Just faced `Invalid url encoding` error and realized that I have to 

`${encodedSourceURL}.${extension}` for encoded URLs

**but**

`${encodedSourceURL}@${extension}` for plain
